src/lib/input/input.ts

```diff
@@ -43,6 +43,15 @@ const MAT_INPUT_INVALID_TYPES = [
 
 let nextUniqueId = 0;
 
+// Boilerplate for applying mixins to MatInput.
+/** @docs-private */
+export class MatInputBase {
+  constructor(public _defaultErrorStateMatcher: ErrorStateMatcher,
+              @Optional() public _parentForm: NgForm,
+              @Optional() public _parentFormGroup: FormGroupDirective,
+              @Optional() @Self() public ngControl: NgControl) {}
```

Comment:
Are the decorators actually needed here? AFAIK the metadata extractor will ignore them since the base class for the `@Component` is the result of a function call